### DISCUSSION TITLE
Setting integer type within ArrayBuilder

### DISCUSSION THREAD

**Discussion by TiniTinyTerminator** [scikit-hep/awkward#3317] [Q&A]
### Description of new feature

It seems that the arraybuilder always assumes int64 when setting up the data. It would be nice to set the integer type like uint8 or uint64. This could help reduce memory load when building huge datasets that only contain smaller numbers. the same goes for floats for example

**Comment by jpivarski** [comment]
That's true about ArrayBuilder, but ArrayBuilder is only for ingesting data from untyped sources, such as Python objects or JSON. Those integers don't have any fixed-width type, like int32 or int64.

There are a few functions for setting the type of Awkward Array, after it has been built from an ArrayBuilder. [ak.values_astype](https://awkward-array.org/doc/main/reference/generated/ak.values_astype.html) sets the numeric type of all numbers in an array, wherever they are, and [ak.enforce_type](https://awkward-array.org/doc/main/reference/generated/ak.enforce_type.html) provides more control by forcing a whole array to fit a given type description. (You can get the type from one array, either as a [ak.types.Type](https://awkward-array.org/doc/main/reference/generated/ak.types.Type.html) object or as a string, change the int64 to something else, and then enforce-type.)

I think it makes sense to change the type after the fact because the arrays of Python objects or JSON can't be too big if you're able to process them at all. And if you're coming from NumPy arrays [ak.from_numpy](https://awkward-array.org/doc/main/reference/generated/ak.from_numpy.html), the `dtype` is preserved.)

Is it okay if I turn this from an issue/feature request into a Discussion?
